src/extension.ts

```diff
@@ -0,0 +1,152 @@
+			extension.setEnabled(true);
+		}),
+
+		vscode.commands.registerCommand('overReactFormatOnSave.disableOverReactFormatOnSave', () => {
+			extension.setEnabled(false);
+		}),
+
+		vscode.workspace.onDidSaveTextDocument((document: vscode.TextDocument) => {
+				extension.onDocumentSave(document);
+		})
+	);
+
+	return extension;
+}
+
+class RunFormatOnSave {
+	private context: vscode.ExtensionContext;
+	private config!: vscode.WorkspaceConfiguration;
+	private channel: vscode.OutputChannel = vscode.window.createOutputChannel('OverReact Format On Save');
+	private minOverReactFormatVersion = '>=3.1.0';
+	private overReactFormatKey = 'over_react_format';
+	private projectDir = vscode.workspace.workspaceFolders ? vscode.workspace.workspaceFolders[0].uri.path : "";
+	private useOverReactFormat:Boolean = false; 
+
+	constructor(context: vscode.ExtensionContext) {
+		this.context = context;
+		this.loadConfig();
+		this.showEnablingChannelMessage();
+
+		if (existsSync(`${this.projectDir}/pubspec.yaml`)) {
+			const pubspecContainsOverReactFormat = devDependenciesContains(this.overReactFormatKey, readFileSync(`${this.projectDir}/pubspec.yaml`, 'utf-8'));
+			const overReactFormatRangeIsValid = dependencyHasValidMinVersion(this.overReactFormatKey, this.minOverReactFormatVersion, readFileSync(`${this.projectDir}/pubspec.yaml`, 'utf-8'), true);
+
+			if (pubspecContainsOverReactFormat && !overReactFormatRangeIsValid) {
+				this.showChannelMessage(`OverReact Format range is not acceptable. Bump the minimum to 3.1.0 to use OverReact Format on Save.`);
+			}
+			this.useOverReactFormat =  pubspecContainsOverReactFormat && overReactFormatRangeIsValid;
+		}
+		// No else condition because there's no penalty for the project not being a Dart project.
+		// The `onDocumentSave` command will just be short-circuited if it is run on non-Dart files.
+	}
+
+	buildCommand(fileName: string) {
+		let command:string;
+
+		const customLineLength = this.config.get<Number>('customLineLength', 0);
+		const shouldDetectLineLength = this.config.get<Boolean>('detectCustomLineLength');
+		const shouldUseCustomLineLength = customLineLength > 0;
+
+		if (shouldUseCustomLineLength && shouldDetectLineLength) {
+			this.showChannelMessage(`Both a custom line-length value and detectCustomLineLength set to true. Skipping line-length detection.`);
+		}
+
+		if (this.useOverReactFormat) {
+			if (shouldUseCustomLineLength) {
+				command =  `pub run over_react_format ${fileName} -l ${customLineLength}`;
+			} else {
+				const detectLineLengthFlag = shouldDetectLineLength && !shouldUseCustomLineLength ? "--detect-line-length" : "";
+				command =  `pub run over_react_format ${fileName} -p ${this.projectDir} ${detectLineLengthFlag}`;
+			} 
+		} else {
+			// TODO add logic to detect line-length from dart_dev's config.dart
+			command = `dartfmt -w ${fileName} ${shouldUseCustomLineLength ? `-l ${customLineLength}` : ''}`;
```

Comment:
What happens if the filename has a semicolon, ampersand, etc. in it?

Reply:
When testing, attempt a file called `what;touch /tmp/vuln;ever.dart`

Reply:
Confirmed RCE - saving a file with that filename resulted in the file `vuln` getting created in the project home.

Reply:
So far I have only been able to exploit this when explicitly opening the maliciously named file and saving it. I am experimenting to see if I can make this happen more covertly before giving this issue a severity rating.